text/0000-module-unification.md

```diff
@@ -0,0 +1,822 @@
+4. Subtrees should be relocatable. If you move a directory to a new place in the tree, its internal structure should all still work.
+5. There should be no cognitive overhead when adding a new thing. The right way should be obvious and not impose unnecessary decisions.
+6. We need clean separation between the namespace of the user's own components, helpers, routes, etc and the framework's own type names ("component", "helper", etc) so that we can disambiguate them and add future ones.
+7. Ideally we will have a place to put tests and styles alongside corresponding components.
+8. Local relative lookup for components and helpers needs to work.
+9. Avoid the "titlebar problem", in which many files are all named "component.js" and you can't tell them apart in your editor.
+10. The resolver should be configured via declarative rules, not imperative
+    JavaScript. In addition to enforcing consistency, this allows addons to
+    augment the system with their own types in a predictable way.
+11. Module structures must be statically analyzable at build time to enable
+    efficiency optimizations.
+12. Module classifications must be extensible and allow for customizations by
+    apps, engines, and addons.
+
+> Note: Constraints > 9 were added based on discussions subsequent to the
+initial meeting.
+
+This proposal attempts to address these constraints with a single consistent
+approach to modules that will make Ember easier to use and learn _and_ improve
+the efficiency of Ember's resolver at run-time.
+
+# Detailed Design
+
+This proposal introduces a new top-level directory, `src`, and establishes
+conventions for organizing modules within it. Also proposed is a refactor of the
+Ember resolver to enable efficient and flexible resolutions based upon the new
+module conventions.
+
+The `src` directory will be used to contain the core ES modules within an Ember
+CLI project, whether that project contains an application, addon, or engine. To
+maintain backward compatibility, the `src` directory will be allowed to
+co-exist alongside existing `app` and/or `addon` directories, although these
+directories should eventually be deprecated.
+
+## Examples
+
+Let's start by taking a look at some examples of Ember projects organized
+according to the proposed conventions.
+
+### Example Application
+
+A simple blogging application could be structured as follows:
+```
+src
+├── data
+│   ├── models
+│   │   ├── comment
+│   │   │   ├── adapter.js
+│   │   │   ├── model.js
+│   │   │   └── serializer.js
+│   │   ├── post
+│   │   │   ├── adapter.js
+│   │   │   ├── model.js
+│   │   │   └── serializer.js
+│   │   └── author.js
+│   └── transforms
+│       └── date.js
+├── init
+│   ├── initializers
+│   │   └── i18n.js
+│   └── instance-initializers
+│       └── auth.js
+├── services
+│   └── auth.js
+├── ui
+│   ├── components
+│   │   ├── date-picker
+│   │   │   ├── component.js
```

Comment:
@rmharrison That's an interesting suggestion. As per the RFC's current state, `components/date-picker/date-picker.js` would not be equivalent to `components/date-picker/component.js`. Instead, `components/date-picker/date-picker.js` would represent a local component available only to `components/date-picker/component.js` (i.e. a date-picker that could be invoked from within another date-picker).

The team discussed the titlebar problem and came to the resolution [mentioned above](https://github.com/emberjs/rfcs/pull/143#issuecomment-250599099):

> We consider the titlebar issue (i.e. multiple files with the same name but in different directories) to be largely an editor-specific problem. We will write a quest issue to ensure that settings are available in common editors to mitigate this problem as necessary.

I will still raise your suggestion with the rest of the team and let you know if we decide to tweak the RFC at all. Thanks for bringing this up!